rust/src/index/ann/flat.rs

```diff
@@ -0,0 +1,122 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use arrow_array::cast::downcast_array;
+use arrow_array::{
+    cast::{as_primitive_array, as_struct_array},
+    RecordBatch,
+};
+use arrow_array::{FixedSizeListArray, Float32Array, StructArray};
+use arrow_ord::sort::sort_to_indices;
+use arrow_schema::{DataType, Field as ArrowField, Schema as ArrowSchema};
+use arrow_select::{concat::concat_batches, take::take};
+use futures::stream::StreamExt;
+
+use super::distance::euclidean_distance;
+use super::SearchParams;
+use crate::arrow::RecordBatchExt;
+use crate::dataset::Dataset;
+use crate::Result;
+
+/// Flat Vector Index.
+///
+/// Flat index is a meta index, that does not build extra index structure,
+/// and uses the full scan.
+///
+/// Reference:
+///   - <https://github.com/facebookresearch/faiss/wiki/Faiss-indexes>
+pub struct FlatIndex<'a> {
+    dataset: &'a Dataset,
+
+    /// Vector column to search for.
+    column: String,
+}
+
+impl<'a> FlatIndex<'a> {
+    /// Create the flat index.
+    pub fn new(dataset: &'a Dataset, column: String) -> Self {
+        Self { dataset, column }
+    }
+
+    /// Search the flat vector index.
+    ///
+    /// Returns a [RecordBatch] with Schema of:
+    ///
+    /// ```
+    /// use arrow_schema::{Schema, Field, DataType};
+    ///
+    /// Schema::new(vec![
+    ///   Field::new("_rowid", DataType::UInt64, false),
+    ///   Field::new("score", DataType::Float32, false),
+    /// ]);
+    /// ```
+    ///
+    /// WARNINGS:
+    ///  - Only supports f32 now. add f64 later.
+    pub async fn search(&self, params: &SearchParams) -> Result<RecordBatch> {
+        let stream = self
+            .dataset
+            .scan()
+            .project(&[&self.column])?
+            .with_row_id()
+            .into_stream();
+
+        let schema = Arc::new(ArrowSchema::new(vec![
+            ArrowField::new("_rowid", DataType::UInt64, false),
+            ArrowField::new("score", DataType::Float32, false),
+        ]));
+        let key_arr: &Float32Array = as_primitive_array(&params.key);
+        let all_scores = stream
+            .map(|b| async {
+                if let Ok(batch) = b {
+                    let key_arr = key_arr.clone();
+                    let value_arr = batch.column_with_name(&self.column).unwrap().clone();
+                    let scores = tokio::task::spawn_blocking(move || {
```

Comment:
Rayon to parallelize dist computation inside the spawn_blocking block?

https://docs.rs/rayon/latest/rayon/iter/trait.IntoParallelRefIterator.html#tymethod.par_iter

Reply:
It should be fine? Given this computation is done on each row in one async function. There is enough parallelism on the record level.